### PR TITLE
Fix rustc panic on Windows due to how stdout/stderr are set up in the process wrapper

### PR DESCRIPTION
**Pull request by ddeville**
I noticed that rustc was panicking with an unexpected null handle when running a `rust_clippy` target on Windows (see below for panic).

I remember that we were specifically setting a `--stderr-file` or `--stdout-file` and I looked at how the process wrapper was actually setting these when invoking rustc/clippy-driver. We are altering the `STARTUPINFO` struct to pass specific file handles for whatever stream we are interested in (see https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/ns-processthreadsapi-startupinfoa for docs).

Setting `hStdInput`, `hStdOutput` or `hStdError` requires setting `STARTF_USESTDHANDLES` on the `dwFlag`. However, the docs say that while setting this flag, *all* three streams should be set on the `STARTUPINFO` to a handle or `INVALID_HANDLE_VALUE`.
Since we were failing to do so, rustc was trying to retrieve `stdout`/`stderr` and tripping on this [assert](https://github.com/rust-lang/rust/blob/18bc4bee9710b181b440a472635150f0d6257713/library/std/src/os/windows/io/handle.rs#L183) since they weren't actually set to anything.

Note that I ran into this with Clippy because it's the first thing I tried that was setting an output file but I suspect it could happen with other configs as well.

```
thread 'main' panicked at 'assertion failed: !handle.is_null()', /rustc/18bc4bee9710b181b440a472635150f0d6257713\library\std\src\os\windows\io\handle.rs:183:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

thread 'main' panicked at 'assertion failed: !handle.is_null()', /rustc/18bc4bee9710b181b440a472635150f0d6257713\library\std\src\os\windows\io\handle.rs:183:9
stack backtrace:
   0:     0x7ff824178c6f - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h4287336b592e4e30
   1:     0x7ff8241a2eba - core::fmt::write::h8a2c40ddb66ccc71
   2:     0x7ff82416b5b8 - <std::io::IoSlice as core::fmt::Debug>::fmt::h37489b9473b33977
   3:     0x7ff82417c1b6 - std::panicking::take_hook::hf14c76592f73c762
   4:     0x7ff82417bc9c - std::panicking::take_hook::hf14c76592f73c762
   5:     0x7ff60d73c2de - <unknown>
   6:     0x7ff82417cac9 - std::panicking::rust_panic_with_hook::hbc0e9c80ca88eac0
   7:     0x7ff82417c53f - rust_begin_unwind
   8:     0x7ff8241795a7 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h4287336b592e4e30
   9:     0x7ff82417c4c9 - rust_begin_unwind
  10:     0x7ff8241d7460 - core::panicking::panic_fmt::hc1b1ca620e7a2c9f
  11:     0x7ff8241d73ac - core::panicking::panic::hfef5cee9afdef02a
  12:     0x7ff8199f250c - <winapi_util[f773672ce88e7109]::console::ScreenBufferInfo>::window_rect
  13:     0x7ff8199f25ef - <winapi_util[f773672ce88e7109]::console::Console>::stdout
  14:     0x7ff8199ee322 - <termcolor[bad27d1885d957ed]::StandardStream>::stderr
  15:     0x7ff8199b5f27 - <rustc_errors[a61263f7b91ac297]::emitter::EmitterWriter>::stderr
  16:     0x7ff60d73c339 - <unknown>
  17:     0x7ff82417cac9 - std::panicking::rust_panic_with_hook::hbc0e9c80ca88eac0
  18:     0x7ff82417c53f - rust_begin_unwind
  19:     0x7ff8241795a7 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h4287336b592e4e30
  20:     0x7ff82417c4c9 - rust_begin_unwind
  21:     0x7ff8241d7460 - core::panicking::panic_fmt::hc1b1ca620e7a2c9f
  22:     0x7ff8241d73ac - core::panicking::panic::hfef5cee9afdef02a
  23:     0x7ff8199f250c - <winapi_util[f773672ce88e7109]::console::ScreenBufferInfo>::window_rect
  24:     0x7ff8199f25ef - <winapi_util[f773672ce88e7109]::console::Console>::stdout
  25:     0x7ff8199ee322 - <termcolor[bad27d1885d957ed]::StandardStream>::stderr
  26:     0x7ff8199d93ca - <rustc_errors[a61263f7b91ac297]::Handler>::with_tty_emitter_and_flags
  27:     0x7ff8199d9361 - <rustc_errors[a61263f7b91ac297]::Handler>::with_tty_emitter
  28:     0x7ff8199161f1 - <rustc_session[ee853a0d40990e94]::parse::ParseSess>::with_silent_emitter
  29:     0x7ff81534443c - <rustc_hir[19483fb96cca7fbf]::definitions::Definitions>::get_root_def
  30:     0x7ff815400efe - <rustc_interface[4e115dbf7cfc9743]::passes::boxed_resolver::BoxedResolver>::to_resolver_outputs
  31:     0x7ff81540d774 - rustc_interface[4e115dbf7cfc9743]::interface::parse_cfgspecs
  32:     0x7ff8151eb83b - <rustc_driver[29a5c9a0ba568b]::RunCompiler>::run
  33:     0x7ff60d733f2c - <unknown>
  34:     0x7ff60d731387 - <unknown>
  35:     0x7ff60d73cdaa - <unknown>
  36:     0x7ff60d7315a6 - <unknown>
  37:     0x7ff60d73299c - <unknown>
  38:     0x7ff8241784d7 - std::rt::lang_start_internal::h1ae7afc8cb06dd46
  39:     0x7ff60d73cde7 - <unknown>
  40:     0x7ff60db2f8c8 - <unknown>
  41:     0x7ff8627d7034 - BaseThreadInitThunk
  42:     0x7ff864362651 - RtlUserThreadStart

thread panicked while processing panic. aborting.
```

### CODE REVIEW

**Comment by ddeville** [comment]
cc @jfgoog

**Review by hlopko — APPROVED** [review]
Looks good to me, but I'm not a Windows person. Please also wait for @UebelAndre's review.

**Comment by ddeville** [comment]
I've actually just noticed a change in behavior: when not setting either a stdout or stderr file we will now set all 3 handles to invalid rather than let the subprocess inherit whatever (console) ones it has. This is incorrect, I'll update.

**Comment by UebelAndre** [comment]
I hope I didn't do anything to portray myself as any kind of windows expert lol but I can take a look after the new changes are pushed 😃

**Comment by dfreese** [comment]
@illicitonion any windows opinions?  Lol

**Comment by ddeville** [comment]
Fwiw I'm also definitely not a Windows person but the workspace I'm attempting to migrate to rules_rust happens to build/run on Linux, MacOS and Windows 😉.

So far I've got pretty much everything ported over and working on Linux and MacOS and I'm slowly making my way through the various remaining Windows issues. I don't know the extent of the work here, some issues like this one were somewhat easy to diagnose, others not so much (I have a crate that chokes on a `#[macro_use]` on an `extern crate ...` with an `ERROR_NOT_ENOUGH_MEMORY` from rustc. If I change it to a `use` – it's using the 2018 edition – I get dozens of `import resolution is stuck, try simplifying other imports`. This only happens on Windows, go figure)

Anyway, trying to put this PR in context of #838, ideally I'd like to get to a point where my workspace can build on Windows with rules_rust. I can't promise anything but for now, I guess I'll go ahead and send PRs whenever I run into issues that I can properly diagnose and fix.

We also happen to have existing rules for Rust internally (forked from a fairly old version of rules_rust if I remember correctly, initially for Linux only but now targeting MacOS and Windows too, with pretty ad-hoc logic unfortunately) that we’ve made work on Windows and have been using to build our desktop client for the last ~2 years. Hopefully I can extract some wisdom from them to upstream here 😄.

**Comment by UebelAndre** [comment]
> Fwiw I'm also definitely not a Windows person but the workspace I'm attempting to migrate to rules_rust happens to build/run on Linux, MacOS and Windows 😉.

hehe, I'm in a similar boat. I've done work in the past when I had to for cross platform compatibility. I wouldn't want anyone to think I'm an expert here but I do think it's important to try and make the rules work on all platforms. So all I can do is try my best and work with my limited experience 😅 

> Anyway, trying to put this PR in context of #838, ideally I'd like to get to a point where my workspace can build on Windows with rules_rust. I can't promise anything but for now, I guess I'll go ahead and send PRs whenever I run into issues that I can properly diagnose and fix.

That sounds awesome! I'd like for you to be able to build your workspace as well! If there are untracked issues, maybe open a new issue for them and link them to #838? I know that'd definitely help me.

Is this PR ready for another review?

**Comment by ddeville** [comment]
Yes it is!

File: util/process_wrapper/system_windows.cc [hunk 1]
```diff
@@ -211,18 +203,42 @@ int System::Exec(const System::StrType& executable,
...
   OutputPipe stderr_pipe;
-  if (!stderr_file.empty() &&
-      !stderr_pipe.CreateEnds(startup_info, /*err*/ true)) {
-    std::cerr << "process wrapper error: failed to create stderr pipe: "
-              << GetLastErrorAsStr();
-    return -1;
+
+  if (!stdout_file.empty() || !stderr_file.empty()) {
+    // We will be setting our own stdout/stderr handles.
+    // Note that when setting `STARTF_USESTDHANDLES` it is critical to set *all* handles or the
```
Comment (UebelAndre) [inline util/process_wrapper/system_windows.cc:210]: Nit: Can we keep the comment lengths more uniform?

**Review by UebelAndre — APPROVED** [review]
Ok, took some spelunking through the Microsoft docs but I think I understand what's happening now. Looks good to me!

**Pull request merged**